src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java

```diff
@@ -0,0 +1,146 @@
+package io.vertx.ext.mongo.impl;
+
+import com.mongodb.async.SingleResultCallback;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.streams.WriteStream;
+import io.vertx.ext.mongo.GridFSInputStream;
+import io.vertx.ext.mongo.util.CircularByteBuffer;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Implementation of {@link GridFSInputStream} which allows streaming with
+ * Vertx's  {@link io.vertx.core.streams.Pump}
+ * <p>
+ * Adapted from com.github.sth.vertx.mongo.streams.GridFSInputStreamImpl
+ *
+ * @author <a href="https://github.com/st-h">Steve Hummingbird</a>
+ */
+public class GridFSInputStreamImpl implements GridFSInputStream {
+
+  public static int DEFAULT_BUFFER_SIZE = 8192;
+  private int writeQueueMaxSize;
+  private final CircularByteBuffer buffer;
+  private Handler<Void> drainHandler;
+  private volatile boolean closed = false;
+  private SingleResultCallback<Integer> pendingCallback;
+  private ByteBuffer outputBuffer;
+
+  public GridFSInputStreamImpl() {
+    buffer = new CircularByteBuffer(DEFAULT_BUFFER_SIZE);
+    writeQueueMaxSize = buffer.capacity();
+  }
+
+  public GridFSInputStreamImpl(final int queueSize) {
+    buffer = new CircularByteBuffer(queueSize);
+    writeQueueMaxSize = queueSize;
+  }
+
+  public void read(ByteBuffer b, SingleResultCallback<Integer> c) {
+    synchronized (buffer) {
```

Comment:
@johnoliver Can you think about the saneness of the concurrency here - it's fine by me but I'm o longer a practioner